### PR TITLE
Add remove_chunk() to Region and tests

### PR DESCRIPTION
**Pull request by icrayix**
Unfortunately I did not find a way to truncate the buffer without having access to an underlying `File` or so.
Maybe add an `impl<File> Region` so we can use [`set_len`](https://doc.rust-lang.org/std/fs/struct.File.html#method.set_len)?

Note: I derived `Clone` on Region as the test `deleted_chunk_at_end_of_buffer_truncates_buffer` uses `into_inner` which consumes self. Please tell me if there's any better way to do so

PS: I'm pretty new to Rust so please don't hesitate to correct mistakes and improve the code :)

### CODE REVIEW

**Comment by owengage** [comment]
Hey, thanks for the contribution. I'll give this a look when I get chance.

With the file issue, we can probably

* create a trait `FileLike` for the `set_len` method 
* implement it for `File`
* have an impl for region where the generic type S implements `FileLike`. 

This might cause some issues though with conflicting impls.

**Comment by owengage** [comment]
It looks to me like you can do this:

```rust
impl<S> Region<S> where S: FileLike {
    pub fn remove_chunk(&mut self, x: usize, z: usize) -> Result<()> {
        todo!()
    }
}

/// FileLike is a sealed trait (users cannot implement it for their types). This
/// allows us to add methods to this trait without breaking backwards
/// compatibility.
pub trait FileLike: private::Sealed {
    fn set_len(&self, size: u64) -> io::Result<()>;
}

impl FileLike for File {
    #[inline(always)]
    fn set_len(&self, size: u64) -> io::Result<()> {
        self.set_len(size)
    }
}

pub(crate) mod private {
    use std::fs::File;

    pub trait Sealed {}
    impl Sealed for File {}
}
```

Then because the `Sealed` trait is effectively `pub(crate)`, you can create a fake file type to do easier unit testing:

```rust
struct FakeFile;

impl crate::region::private::Sealed for FakeFile {}

impl FileLike for FakeFile {
    fn set_len(&self, size: u64) -> std::io::Result<()> {
        todo!()
    }
}
```

It might be a bit of a pain to also implement `Read`, `Write`, and `Seek` though.

**Comment by icrayix** [comment]
I have to admit that that's a pretty neat solution :)

The only question remaining for me is whether we should have remove_chunk for non-file regions as well.
Not sure on how to implement the truncate functionality on FileLike only tho.

**Comment by owengage** [comment]
> The only question remaining for me is whether we should have remove_chunk for non-file regions as well. Not sure on how to implement the truncate functionality on FileLike only tho.

That's a good point. We won't be able to have a `remove_chunk` method on both `impl` blocks as they will conflict.

I think maybe `remove_chunk` should be implemented on the existing `Read+Write+Seek` impl, and a separate `truncate_file` method on the `File` impl that purely checks to last chunk on disk and calls `set_len` if required.

The documentation for both methods should reference each other to let people know of the potential pitfall with files.

**Comment by owengage** [comment]
Thinking about it some more, I'm wondering if `Region` should instead expose the logical end of the stream, eg via a `len` method that returns the number of bytes taken up by the region data.

This means that if you were working with a `File`, you would do something like...

```rust
let region = Region::new(f);
// manipulate region a bunch
let len = region.len();
let f = region.into_inner();
f.set_len(len);
```

This would mean that other types like `File` and follow a similar pattern for their specific type.

**Comment by icrayix** [comment]
But doesn't into_inner consume self?
One wouldn't be able to continue using the region, but rather create a new one, right?

**Comment by icrayix** [comment]
Other than that I really like that approach better as the exposed api is a lot more consistent.

**Comment by owengage** [comment]
Yep, you've got it. The `FileLike` definitely seems ugly in comparison.

Be aware that Minecraft likes chunks to be padded to a sector length (see #75), so we don't want to truncate the file to the exact byte-end of the chunk data.

File: fastanvil/src/region.rs [hunk 1]
```diff
@@ -25,6 +26,7 @@ pub(crate) const CHUNK_HEADER_SIZE: usize = 5;
 /// File). This does not concern itself with manipulating chunk data, users are
 /// expected to use `fastnbt` or other deserialization method to manipulate the
 /// chunk data itself.
+#[derive(Clone)]
```
Comment (owengage) [inline fastanvil/src/region.rs:29]: I was suprised this worked given `S` isn't marked as `Clone`, but turns out the derived `Clone` adds a bound for `S`, so `Region` won't implement it for eg `S = File` but will for `S = Cursor<Vec<u8>>`. I'm happy with this.

**Comment by icrayix** [comment]
Here we go.
A lot more verbose than expected but couldn't find any better way.

The function doc could perhaps be improved but my english skills are lacking for that.

**Comment by icrayix** [comment]
Like that way better than the solution before :)

**Comment by owengage** [comment]
This looks good. I'll give it a closer look when I get chance. Thanks again! :)

**Comment by icrayix** [comment]
Hi,

have you already taken a look into it?
Not willing to annoy you just wanted to make sure you still have that on your radar :)

**Comment by owengage** [comment]
I have not, sorry! I had a quick look over now and it looks good, so I'm going to go ahead and merge. Thank you :)

File: fastanvil/src/region.rs [hunk 2]
```diff
@@ -149,9 +150,45 @@ where
...
+    /// # use fastanvil::Region;
+    /// # use fastanvil::Result;
+    /// # use std::io::Seek;
+    /// # fn main() -> Result<()> {
+    /// let mut file = std::fs::File::open("foo.mca")?;
+    /// let mut region = Region::from_stream(file)?;
+    /// // manipulate region
+    /// let mut file = region.into_inner()?;
+    /// let len = file.stream_position()?;
+    /// file.truncate(len)?;
```
Comment (owengage) [inline fastanvil/src/region.rs:169]:
```suggestion
    /// file.truncate(len)?;
    file.set_len(len)?;
```

**Comment by owengage** [comment]
Okay, tried to fix a build issue with the suggest feature and I made a mess of it!

There's a build failure that needs fixing, basicallly just commenting a line of code in the doccomment where `cargo test` is failing. If you change that I can merge. :)

**Comment by icrayix** [comment]
Ooops.

No clue on why I used file.truncate there even though that method doesn't even exist lol.

I also addressed a clippy warning there.

**Pull request merged**